##Manipulating and filtering ASV table##

setwd("/Users/marianacamaradosreis/Documents/PhD/Chapter1/Final_results/analysis_R/200331_microbiome_analysis")

##to install new necessary packages:  BiocManager::install("name of package")"


library(phyloseq); packageVersion("phyloseq")
library(microbiome); packageVersion("microbiome")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(data.table); packageVersion("data.table")
library(vegan); packageVersion("vegan")
library(pheatmap); packageVersion("pheatmap")
library(microbiome); packageVersion("microbiome")
library(dplyr); packageVersion("dplyr")
library(tidyverse); packageVersion("tidyverse")
library(RColorBrewer); packageVersion("RColorBrewer")
library(dada2) ;packageVersion("dada2")
library("DESeq2");packageVersion("DESeq2")
library(readr);packageVersion("readr")
library(ape);packageVersion("ape")
library(taxa);packageVersion("taxa")
library("compositions");packageVersion("taxa")
library(metacoder); packageVersion("metacoder")
library("GUniFrac")
theme_set(theme_bw()) ##define a default theme for ggplot graphics

####Importing tables and performing some summary##

out <- readRDS("200401_out.rds")
seqtab <- readRDS("200406_seqtab.rds")
seqtab_nochimera <- readRDS("200407_seqtab_nochimera.rds")
seqtab_nochimera_pooled <- readRDS("200619_seqtab_nochimera_pooled.rds")

sum(seqtab)
sum(seqtab_nochimera)
sum(seqtab_nochimera_pooled)

sum(seqtab_nochimera_pooled)/sum(seqtab)

dim(seqtab)
dim(seqtab_nochimera)
dim(seqtab_nochimera_pooled)
dim(seqtab_nochimera_pooled)/dim(seqtab)

table(nchar(getSequences(seqtab)))
table(nchar(getSequences(seqtab_nochimera_pooled)))

###ASVs per sequence legth###

table <- as.data.frame(table(nchar(colnames(seqtab_nochimera_pooled))))
colnames(table) <- c("LENGTH","COUNT")

ggplot(table,aes(x=LENGTH,y=COUNT)) + 
  geom_histogram(stat="identity") + 
  ggtitle("Sequence Lengths by SEQ Count") +
  theme_bw() +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5,size=10)) +
  theme(axis.text.y=element_text(size=10))

###Reads per sequence length###

table2 <- tapply(colSums(seqtab_nochimera_pooled), nchar(colnames(seqtab_nochimera_pooled)), sum)
table2 <- data.frame(key=names(table2), value=table2)

colnames(table2) <- c("LENGTH","ABUNDANCE")

ggplot(table2,aes(x=LENGTH,y=ABUNDANCE)) + 
  geom_histogram(stat="identity") + 
  ggtitle("Sequence Lengths by SEQ Abundance") +
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5,size=10)) +
  theme(axis.text.y=element_text(size=10))

#restricting sequence legth since I have a high variability

tapply(colSums(seqtab_nochimera_pooled), nchar(colnames(seqtab_nochimera_pooled)), sum)
seqlens <- nchar(colnames(seqtab_nochimera_pooled))
seqtab.filt <- seqtab_nochimera_pooled[, (seqlens <= 377) & (seqlens >= 365)]
sum_filt <- tapply(colSums(seqtab.filt), nchar(colnames(seqtab.filt)), sum)

##Obtaining managable names for ASVs

asv_seqs <- colnames(seqtab.filt)
asv_headers <- vector(dim(seqtab.filt)[2], mode="character")

for (i in 1:dim(seqtab.filt)[2]) {
  asv_headers[i] <- paste(">ASV", i, sep="_")
}

# making and writing out a fasta of the final ASV seqs:

asv_fasta <- c(rbind(asv_headers, asv_seqs))
write(asv_fasta, "200621_ASVs_ehux_restricted_chimera_pooled.fa")


# count table:
asv_tab <- t(seqtab.filt)
row.names(asv_tab) <- sub(">", "", asv_headers)
colnames(asv_tab) <- gsub(colnames(asv_tab), pattern = "\\-", replacement = "_") ##replacing "-" by "_" samples names
write.table(asv_tab, "200621_ASVs_counts_ehux_restricted_chimera_pooled.tsv", sep="\t", quote=F, col.names=NA)
asv_tab
sum(asv_tab)


###Manipulating tha data using tidyverse####

otu_data <- read_tsv("/Users/marianacamaradosreis/Documents/PhD/Chapter3/Final_results/200620_ASVs_counts_microbiome_restricted_chimera_pooled.tsv")
print(otu_data)

tax_data <- read_tsv("/Users/marianacamaradosreis/Documents/PhD/Chapter3/Final_results/200621_ASVs_microbiome_restricted_chimera_pooled_vsearch.txt") %>%
  replace(., is.na(.), "unknown") ##using taxonomy generated by vsearch
print(tax_data)

metadata <- read_tsv("/Users/marianacamaradosreis/Documents/PhD/Chapter3/Final_results/200505_metadata_simplified.txt")
print(metadata)

colnames(otu_data)[2:409] == metadata$id
otu_data$ASVId == tax_data$ASVId

otu_data <- left_join(otu_data, tax_data,
                      by = c("X1" = "ASV_Id"))

tail(colnames(otu_data), n = 10)
print(otu_data$Dom)


##Filtering out Chloroplasts, Mitochondria and Eukaryotes###
####empty cells need to be replaced otherwise they are discarded

otu_data2 <- otu_data %>% 
  filter(., Dom == 'Bacteria' & Order != 'Chloroplast' & Family != 'Mitochondria')
dim(otu_data2)
#write.table(otu_data2, "200507_table_nochloroplast.txt", sep="\t")

##Keeping only the samples that we will include in the analysis##

pivoted <- otu_data2 %>%
  rownames_to_column %>% 
  gather(var, value, -rowname) %>% 
  pivot_wider(names_from=rowname, values_from=value)

colnames(pivoted) <- pivoted[1,]
pivoted <- pivoted[-1,,drop=F]

##Taking only ehux data###

otu_data_long <-  otu_data %>% 
  gather(var, value, -X1) %>% inner_join(., metadata, by = c("var" = "id")) %>% 
  filter(keep_first_analysis == "yes") %>%  mutate_at(vars(value), list(as.numeric))

filtered <- otu_data_long %>%
  mutate(Total = sum(value)) %>%
  group_by(X1) %>% 
  mutate(ASV_abund = sum(value)) %>% 
  filter(ASV_abund > 0)

sum(filtered$value)

filt_wide <- filtered %>%
  select(X1, value, var) %>% 
  pivot_wider(names_from = X1, values_from = value) %>% 
  replace(., is.na(.), 0) %>% column_to_rownames(., var = "var")
dim(filt_wide)
sum(filt_wide)


###Creating distance matrix and plotting HCA of replicates to see if the replicates of same samples appear close to each other

otu_data_long <-  otu_data2 %>% 
  gather(var, value, -X1) %>% inner_join(., metadata, by = c("var" = "id")) %>% 
  filter(keep_ehux == "yes") %>%  mutate_at(vars(value), list(as.numeric))

filtered <- otu_data_long %>%
  mutate(Total = sum(value)) %>%
  group_by(X1) %>% 
  mutate(ASV_abund = sum(value)) %>% 
  filter(ASV_abund > 0)
min(filtered$ASV_abund)

filt_wide <- filtered %>%
  select(X1, value, var) %>% 
  pivot_wider(names_from = X1, values_from = value) %>% 
  replace(., is.na(.), 0) %>% column_to_rownames(., var = "var") %>% 
  decostand(., method="hellinger")

dim(filt_wide)
sum_asv <- as.matrix(colSums(filt_wide))


dist_b <- vegdist(filt_wide, "euclidean")
hc <- hclust(dist_b, "complete")
plot(hc, cex=0.3)

mds <- metaMDS(filt_wide,  dist="euclidean", try=3000)
mds2 <- metaMDS(filt_wide,  dist="euclidean", try=3000, previous.best = mds)
mds_data <- as.data.frame(mds$points)

mds_data$SampleID <- rownames(mds_data)
mds_data <- left_join(mds_data, metadata, by = c("SampleID" = "id"))


ggplot(mds_data, aes(x = MDS1, y = MDS2, label = rcc)) +
  geom_path(data=mds_data,aes(x=MDS1,y=MDS2,group=rcc, size=1.5))+
  geom_point(size=3) + geom_text()



###Merging samples based on average of present ASVs##
#important to use na.rm=FALSE, it will not do mean if of the replicates has one na
pivoted_met <- inner_join(pivoted, metadata, by = c("ASVId" = "id")) %>% 
  filter(keep_ehux == "yes") %>% na_if(., "0") %>% mutate_at(vars(starts_with("ASV_")),list(as.numeric))

group <- group_by(pivoted_met, rcc)  %>% summarise(across(starts_with("ASV_"), .fns = mean, na.rm=FALSE)) %>% replace(., is.na(.), 0)
#write.table(group, "200528_core_ehux_merged_abund_filt.txt", sep="\t")
